### DISCUSSION TITLE
Bonsai experienced an error :(

### DISCUSSION THREAD

**Discussion by JuveSPb** [IfcOpenShell/IfcOpenShell#5528] [Q&A]
![error](https://github.com/user-attachments/assets/e667bc31-8ad4-4dd2-a9a0-70618928a6a9)

Dear developers,

I have faced a problem with ifc opening. Here is the screenshot and error message. Can you help me with solving the problem?

Best Regards, Mikhail.

```
os: Windows
os_version: 10.0.19045
python_version: 3.11.7
architecture: ('64bit', 'WindowsPE')
machine: AMD64
processor: Intel64 Family 6 Model 140 Stepping 2, GenuineIntel
blender_version: 4.2.1 LTS
bonsai_version: 0.8.0
bonsai_commit_hash: None
last_actions: 
last_error: Traceback (most recent call last):
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\bim\module\project\operator.py", line 751, in finish_loading_project
    bpy.ops.bim.load_project_elements()
  File "C:\Program Files\Blender Foundation\Blender 4.2\4.2\scripts\modules\bpy\ops.py", line 109, in __call__
    ret = _op_call(self.idname_py(), kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Error: Python: Traceback (most recent call last):
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\bim\module\project\operator.py", line 832, in execute
    ifc_importer.execute()
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\bim\import_ifc.py", line 244, in execute
    self.create_elements()
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\bim\import_ifc.py", line 630, in create_elements
    self.create_generic_elements(self.elements)
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\bim\import_ifc.py", line 656, in create_generic_elements
    objects.add(self.create_product(element))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\bim\import_ifc.py", line 896, in create_product
    obj, tool.Loader.apply_blender_offset_to_matrix_world(obj, self.get_element_matrix(element))
                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\bonsai\bim\import_ifc.py", line 1276, in get_element_matrix
    result = ifcopenshell.util.placement.get_local_placement(element.ObjectPlacement)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\ifcopenshell\util\placement.py", line 130, in get_local_placement
    return np.dot(parent, get_axis2placement(placement.RelativePlacement))
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\ifcopenshell\util\placement.py", line 79, in get_axis2placement
    shape = ifcopenshell.geom.create_shape(settings, placement)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\ifcopenshell\geom\main.py", line 411, in create_shape
    ifcopenshell_wrapper.create_shape(
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\4.2\extensions\.local\lib\python3.11\site-packages\ifcopenshell\ifcopenshell_wrapper.py", line 6075, in create_shape
    return _ifcopenshell_wrapper.create_shape(*args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Unexpected topology
Location: C:\Program Files\Blender Foundation\Blender 4.2\4.2\scripts\modules\bpy\ops.py:109


ifc: PKS2.D.P000.5.0UJE99GML&&&.052.DC.0001.S.ifc
schema: IFC4
preprocessor_version: Processor version 6.0.0.0
originating_system: Xbim.IO.MemoryModel
```

**Comment by aothms** [comment]
Given that it's in the ifcopenshell.geom.create_shape(settings, placement) call, it's most likely an invalid IfcAxis2Placement3D. I finally implemented a bit more gracefully handling of this. In the meantime you can look for `IFCAXIS2PLACEMENT3D($` in your model and replace the first `$` with a reference to an `IFCCARTESIANPOINT((0.,0.,0.))` (or wait until we publish a new Bonsai build with updated ifcopenshell). Trying to fix it in the model would be a good way though to verify before hand whether the fix is sufficient for your problem.

You can also run `python -m ifcopenshell.validate` on your model and see if there are references to an error indicating that the Location attribute is not optional. Or upload it to validate.buildingsmart.org.

**Reply by JuveSPb** [reply]
> Given that it's in the ifcopenshell.geom.create_shape(settings, placement) call, it's most likely an invalid IfcAxis2Placement3D. I finally implemented a bit more gracefully handling of this. In the meantime you can look for `IFCAXIS2PLACEMENT3D($` in your model and replace the first `$` with a reference to an `IFCCARTESIANPOINT((0.,0.,0.))` (or wait until we publish a new Bonsai build with updated ifcopenshell). Trying to fix it in the model would be a good way though to verify before hand whether the fix is sufficient for your problem.
> 
> You can also run `python -m ifcopenshell.validate` on your model and see if there are references to an error indicating that the Location attribute is not optional. Or upload it to validate.buildingsmart.org.

Hello Thomas,
I am far from coding and as I understand you correctly, the following actions with ifc-file should be done. Please, check my corrections on the screenshot. After having been done these modifications , I tried to open problem ifc-file again, but the error still occurred.
Thank you for your time and help!

BR, Mikhail

![image](https://github.com/user-attachments/assets/e515a4ae-bf74-48e8-8ea4-4d7e3a17b121)

**Reply by Moult** [reply]
That change looks correct but make sure you change it for all instances in the file.

**Reply by JuveSPb** [reply]
> That change looks correct but make sure you change it for all instances in the file.

I changed it for all instances in the file. After that, I tried to open it again. It looked like Blender has frozen, but in 15-20 minutes it has finally opened it. But there are some problems with geometry in the file. It all became 2D shaped. You can see it on screenshots. The first one is from BIMvision app.
![BIMvision](https://github.com/user-attachments/assets/5b8e46e2-4d8e-4aaa-915b-d999f671a72e)
![Blender](https://github.com/user-attachments/assets/863efdaa-15b3-491a-90f3-d6068ab34789)

**Reply by Moult** [reply]
Can you help share the file? https://forms.gle/ZMs1Wf8j2NLDixQe7 if private.

**Comment by Varcose** [comment]
Hello community, 
After 25 years of using proprietary software I decided to start my journey in the world of BIM development, I was pleased to found the open source development of ifc  very advanced, I hope to be able to contribute more in the future, for now, the problem that I have is that the Bonsai 8.0 extention didn't install in to the Blender 4.2 LTS.
I'm sharing the log and one image, (it has the same scary red notes), I hope this shared information  will be enough to find the way to help me to solve the problem, thank you for the guidance, I would like to start my learning as soon as possible.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/victo/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<link rel=Edit-Time-Data
href="file:///C:/Users/victo/AppData/Local/Temp/msohtmlclip1/01/clip_editdata.mso">

<link rel=themeData
href="file:///C:/Users/victo/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/victo/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">

<style>

</style>

</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>


<p class=MsoNormal>os: Windows<o:p></o:p></p>

<p class=MsoNormal>os_version: 10.0.22631<o:p></o:p></p>

<p class=MsoNormal>python_version: 3.11.7<o:p></o:p></p>

<p class=MsoNormal>architecture: ('64bit', 'WindowsPE')<o:p></o:p></p>

<p class=MsoNormal>machine: AMD64<o:p></o:p></p>

<p class=MsoNormal>processor: Intel64 Family 6 Model 154 Stepping 3,
GenuineIntel<o:p></o:p></p>

<p class=MsoNormal><span lang=FR style='mso-ansi-language:FR'>blender_version:
4.2.2 LTS<o:p></o:p></span></p>

<p class=MsoNormal><span lang=FR style='mso-ansi-language:FR'>bonsai_version:
0.8.0<o:p></o:p></span></p>

<p class=MsoNormal>bonsai_commit_hash: None<o:p></o:p></p>

<p class=MsoNormal>last_actions: <o:p></o:p></p>

<p class=MsoNormal>last_error: Traceback (most recent call last):<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;C:\Users\victo\AppData\Roaming\Blender
Foundation\Blender\4.2\scripts\addons\blenderbim\libs\site\packages\ifcopenshell\ifcopenshell_wrapper.py&quot;,
line 14, in swig_import_helper<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>    </span>return
importlib.import_module(mname)<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>          
</span>^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;C:\Program Files\Blender Foundation\Blender
4.2\4.2\python\Lib\importlib\__init__.py&quot;, line 126, in import_module<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>    </span>return
_bootstrap._gcd_import(name[level:], package, level)<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>          
</span>^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;&lt;frozen importlib._bootstrap&gt;&quot;, line 1204, in _gcd_import<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;&lt;frozen importlib._bootstrap&gt;&quot;, line 1176, in _find_and_load<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;&lt;frozen importlib._bootstrap&gt;&quot;, line 1147, in
_find_and_load_unlocked<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;&lt;frozen importlib._bootstrap&gt;&quot;, line 676, in _load_unlocked<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;&lt;frozen importlib._bootstrap&gt;&quot;, line 573, in module_from_spec<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;&lt;frozen importlib._bootstrap_external&gt;&quot;, line 1233, in
create_module<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;&lt;frozen importlib._bootstrap&gt;&quot;, line 241, in
_call_with_frames_removed<o:p></o:p></p>

<p class=MsoNormal>ImportError: DLL load failed while importing
_ifcopenshell_wrapper: The specified module could not be found.<o:p></o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>During handling of the above exception, another exception
occurred:<o:p></o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>Traceback (most recent call last):<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;C:\Users\victo\AppData\Roaming\Blender
Foundation\Blender\4.2\scripts\addons\blenderbim\libs\site\packages\ifcopenshell\__init__.py&quot;,
line 61, in &lt;module&gt;<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>    </span>from . import
ifcopenshell_wrapper<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;C:\Users\victo\AppData\Roaming\Blender
Foundation\Blender\4.2\scripts\addons\blenderbim\libs\site\packages\ifcopenshell\ifcopenshell_wrapper.py&quot;,
line 17, in &lt;module&gt;<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>   
</span>_ifcopenshell_wrapper = swig_import_helper()<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>                           
</span>^^^^^^^^^^^^^^^^^^^^<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;C:\Users\victo\AppData\Roaming\Blender
Foundation\Blender\4.2\scripts\addons\blenderbim\libs\site\packages\ifcopenshell\ifcopenshell_wrapper.py&quot;,
line 16, in swig_import_helper<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>    </span>return
importlib.import_module('_ifcopenshell_wrapper')<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>          
</span>^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;C:\Program Files\Blender Foundation\Blender
4.2\4.2\python\Lib\importlib\__init__.py&quot;, line 126, in import_module<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>    </span>return
_bootstrap._gcd_import(name[level:], package, level)<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>          
</span>^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^<o:p></o:p></p>

<p class=MsoNormal>ModuleNotFoundError: No module named '_ifcopenshell_wrapper'<o:p></o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>During handling of the above exception, another exception
occurred:<o:p></o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>Traceback (most recent call last):<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;C:\Users\victo\AppData\Roaming\Blender
Foundation\Blender\4.2\extensions\blender_org\bonsai\__init__.py&quot;, line
225, in &lt;module&gt;<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>    </span>import
ifcopenshell.api<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>  </span>File
&quot;C:\Users\victo\AppData\Roaming\Blender
Foundation\Blender\4.2\scripts\addons\blenderbim\libs\site\packages\ifcopenshell\__init__.py&quot;,
line 69, in &lt;module&gt;<o:p></o:p></p>

<p class=MsoNormal><span style='mso-spacerun:yes'>    </span>raise
ImportError(&quot;IfcOpenShell not built for '%s'&quot; % python_distribution)<o:p></o:p></p>

<p class=MsoNormal>ImportError: IfcOpenShell not built for
'windows\64bit\python3.11'<o:p></o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>binary_file_name: _ifcopenshell_wrapper.cp311-win_amd64.pyd<o:p></o:p></p>

<p class=MsoNormal>binary_python_version: 3.11<o:p></o:p></p>

<p class=MsoNormal><![if !vml]><span style='mso-ignore:vglayout'>

</span><![endif]><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>



![image](https://github.com/user-attachments/assets/58cc9637-e20c-4456-9326-a28ae542027d)

**Reply by Moult** [reply]
Can you try following https://docs.bonsaibim.org/quickstart/installation.html with a fresh Blender 4.2 installation? If you still have an error you can join the https://osarch.org/chat live chat and we can screenshare.

**Comment by Varcose** [comment]
Hi Thomas,It works, now I’m ready to start, thank you,  seams a lot to catch up
Víctor Arcos
    On Wednesday, October 9, 2024 at 08:36:30 PM GMT-5, Dion Moult ***@***.***> wrote:  
 
 


Can you try following https://docs.bonsaibim.org/quickstart/installation.html with a fresh Blender 4.2 installation? If you still have an error you can join the https://osarch.org/chat live chat and we can screenshare.
